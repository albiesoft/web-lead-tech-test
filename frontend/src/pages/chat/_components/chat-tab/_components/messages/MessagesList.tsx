import type { FC } from "react";
import MessageItem from "./_components/MessageItem";
import { useGetMessages } from "@/hooks/useGetMessages";
import { differenceInSeconds, format } from "date-fns";
import { daysAgo } from "@/utils/dates";

export const MessagesList: FC = () => {
  const { messages } = useGetMessages();

  return (
    <div className="flex-1 flex flex-col p-[5px] overflow-auto max-h-[490px]">
      <div className="mt-auto">
        {messages.map((message, idx) => {
          const oldMsg = messages[idx - 1];
          if (!!oldMsg) {
            const diff = differenceInSeconds(
              message.timestamp,
              oldMsg.timestamp
            );

            if (
              diff > 3600 ||
              (oldMsg.senderId === message.senderId && diff > 20)
            ) {
              return (
                <>
                  <div>
                    <p className="text-sm text-center mb-6">
                      <strong>{daysAgo(message.timestamp)}</strong>{" "}
                      <span>{format(message.timestamp, "H:mm aa")}</span>
                    </p>
                  </div>
                  <MessageItem message={message} key={message.id} />
                </>
              );
            }
          } else {
            return (
              <>
                <div>
                  <p className="text-sm text-center mb-6">
                    <strong>{daysAgo(message.timestamp)}</strong>{" "}
                    <span>{format(message.timestamp, "H:mm aa")}</span>
                  </p>
                </div>
                <MessageItem message={message} key={message.id} />
              </>
            );
          }
          return <MessageItem message={message} key={message.id} />;
        })}
      </div>
    </div>
  );
};
