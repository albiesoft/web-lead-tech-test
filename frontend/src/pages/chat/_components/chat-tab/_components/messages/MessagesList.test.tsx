import SocketProvider from "@/context/socketContext";
import { MessagesList } from "./MessagesList";
import { render } from "@testing-library/react";
import type { Message } from "@/store/messages.store";

const mockedMessages: Message[] = [
  {
    id: 1,
    senderId: 1,
    recipientId: 2,
    content: "Hello",
    timestamp: new Date().toISOString(),
  },
  {
    id: 2,
    senderId: 1,
    recipientId: 2,
    content: "How are you?",
    timestamp: new Date().toISOString(),
  },
  {
    id: 3,
    senderId: 2,
    recipientId: 1,
    content: "Hey!",
    timestamp: new Date().toISOString(),
  },
  {
    id: 4,
    senderId: 2,
    recipientId: 1,
    content: "I am good and you?",
    timestamp: new Date().toISOString(),
  },
];

vi.mock("@/hooks/useGetMessages", () => ({
  useGetMessages: () => ({ messages: mockedMessages, loading: false }),
}));

describe("<MessageList />", () => {
  const wrapper = () => (
    <SocketProvider>
      <MessagesList />
    </SocketProvider>
  );

  it("should render correctly", () => {
    const comp = render(wrapper());

    expect(comp).toMatchSnapshot();
  });
});
