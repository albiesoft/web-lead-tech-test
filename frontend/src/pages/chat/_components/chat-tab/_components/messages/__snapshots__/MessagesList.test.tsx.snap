// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<MessageList /> > should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex-1 flex flex-col p-[5px] overflow-auto max-h-[490px]"
      >
        <div
          class="mt-auto"
        >
          <div>
            <p
              class="text-sm text-center mb-6"
            >
              <strong>
                Today
              </strong>
               
              <span>
                17:25 PM
              </span>
            </p>
          </div>
          <div
            class="flex flex-row"
          >
            <div>
              <div
                class="rounded-lg px-4 py-3 text-sm m-2 bg-gray-50 rounded-bl-none"
              >
                Hello
              </div>
            </div>
          </div>
          <div
            class="flex flex-row"
          >
            <div>
              <div
                class="rounded-lg px-4 py-3 text-sm m-2 bg-gray-50 rounded-bl-none"
              >
                How are you?
              </div>
            </div>
          </div>
          <div
            class="flex flex-row-reverse"
          >
            <div>
              <div
                class="rounded-lg px-4 py-3 text-sm m-2 bg-accent text-white rounded-br-none"
              >
                Hey!
              </div>
            </div>
          </div>
          <div
            class="flex flex-row-reverse"
          >
            <div>
              <div
                class="rounded-lg px-4 py-3 text-sm m-2 bg-accent text-white rounded-br-none"
              >
                I am good and you?
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex-1 flex flex-col p-[5px] overflow-auto max-h-[490px]"
    >
      <div
        class="mt-auto"
      >
        <div>
          <p
            class="text-sm text-center mb-6"
          >
            <strong>
              Today
            </strong>
             
            <span>
              17:25 PM
            </span>
          </p>
        </div>
        <div
          class="flex flex-row"
        >
          <div>
            <div
              class="rounded-lg px-4 py-3 text-sm m-2 bg-gray-50 rounded-bl-none"
            >
              Hello
            </div>
          </div>
        </div>
        <div
          class="flex flex-row"
        >
          <div>
            <div
              class="rounded-lg px-4 py-3 text-sm m-2 bg-gray-50 rounded-bl-none"
            >
              How are you?
            </div>
          </div>
        </div>
        <div
          class="flex flex-row-reverse"
        >
          <div>
            <div
              class="rounded-lg px-4 py-3 text-sm m-2 bg-accent text-white rounded-br-none"
            >
              Hey!
            </div>
          </div>
        </div>
        <div
          class="flex flex-row-reverse"
        >
          <div>
            <div
              class="rounded-lg px-4 py-3 text-sm m-2 bg-accent text-white rounded-br-none"
            >
              I am good and you?
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
